xc7/utils/prjxray_tile_type_import.py

```diff
@@ -0,0 +1,268 @@
+#!/usr/bin/env python3
+"""
+Import the top level tile information from Project X-Ray database
+files.
+
+Project X-Ray specifies the connections between tiles and the connect
+between tiles and their sites.  prjxray_tile_type_import generates a VPR tile
+that has the correct tile pin location, switchblock location and fc information.
+
+Optionally, if there are available equivalent tiles, the tile information is filled
+with the equivalent tile pin mapping as well. This is needed by VPR to have a correct
+translation of the pin names when using the equivalent tile.
+
+"""
+
+from __future__ import print_function
+import argparse
+import os
+import sys
+import prjxray.db
+import os.path
+import simplejson as json
+
+import lxml.etree as ET
+
+
+def object_ref(pb_name, pin_name, pin_idx=None):
+    pin_addr = ''
+    if pin_idx is not None:
+        pin_addr = '[{}]'.format(pin_idx)
+
+    return '{}.{}{}'.format(pb_name, pin_name, pin_addr)
+
+
+def main():
+    mydir = os.path.dirname(__file__)
+    prjxray_db = os.path.abspath(
+        os.path.join(mydir, "..", "..", "third_party", "prjxray-db")
+    )
+
+    db_types = prjxray.db.get_available_databases(prjxray_db)
+
+    parser = argparse.ArgumentParser(
+        description=__doc__, fromfile_prefix_chars='@', prefix_chars='-~'
+    )
+
+    parser.add_argument(
+        '--part',
+        choices=[os.path.basename(db_type) for db_type in db_types],
+        help="""Project X-Ray database to use."""
+    )
+
+    parser.add_argument('--tile', help="""Tile to generate for""")
+
+    parser.add_argument(
+        '--tiles-directory', help="""Diretory where tiles are defined"""
+    )
+
+    parser.add_argument(
+        '--equivalent-tiles',
+        help="""Comma separated list of equivalent tiles.""",
+    )
+
+    parser.add_argument(
+        '--pin-prefix',
+        help=
+        """Comma separated list of prefix translation pairs for equivalent tiles.""",
+    )
+
+    parser.add_argument(
+        '--output-tile',
+        nargs='?',
+        type=argparse.FileType('w'),
+        default=sys.stdout,
+        help="""File to write the output too."""
+    )
+
+    parser.add_argument(
+        '--pin_assignments', required=True, type=argparse.FileType('r')
+    )
+
+    args = parser.parse_args()
+
+    tile_name = args.tile
+
+    xi_url = "http://www.w3.org/2001/XInclude"
+    ET.register_namespace('xi', xi_url)
+
+    # Macros used to select relevant port names
+    IN_PORT_TAG = ['input', 'clock']
+    OUT_PORT_TAG = ['output']
+
+    ##########################################################################
+    # Utility functions to create tile tag.                                  #
+    ##########################################################################
+
+    def get_port_name(port):
+        """Used to split the port in port full name, port name and pin name"""
+        split = port.split('_')
+        pin = split[-1]
```

Comment:
This will not work in the general case.  Let's just rely on exact pin matches for now?

Reply:
The problem with the exact pin matches is the there are two same pins in each CLB (e.g. `CLBLL_L_A` and `CLBLL_LL_A`) and each of them has to be related to the equivalent pin in the equivalent tile (e.g. `CLBLL_L_A` must be related to `CLBLM_M_A`).
Now, with the tile split being almost complete this will not cause an issue and `get_port_name` could be avoided.

Anyways I need to check if this works in the general case, as you said.

Reply:
I have checked and this does not work in the general case. With the tile split this function can be avoided, as there is not anymore a situation as for the CLBs and we can rely on exact pin matches.